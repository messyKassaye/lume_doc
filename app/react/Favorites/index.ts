export { FavoriteBanner } from './components/FavoriteBanner';
