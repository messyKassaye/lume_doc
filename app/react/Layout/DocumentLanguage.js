import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import getLanguage from 'shared/languagesList';
import t from '../I18N/t';

export class DocumentLanguage extends Component {
  render() {
    const { doc } = this.props;

    if (!doc.get('file')) {
      return null;
    }

    if (doc.get('file')) {
      const fileLanguage = doc.getIn(['file', 'language']);
      if (fileLanguage && fileLanguage !== 'other') {
        if (this.props.locale === getLanguage(fileLanguage, 'ISO639_1')) {
          return null;
        }

        return (
          <span className="item-type__documentLanguage">
            <span>{getLanguage(fileLanguage, 'ISO639_1') || fileLanguage}</span>
          </span>
        );
      }

      return (
        <span className="item-type__documentLanguage">
          <span>{t('System', 'Other')}</span>
        </span>
      );
    }
  }
}

DocumentLanguage.propTypes = {
  doc: PropTypes.object,
  locale: PropTypes.string,
};

export const mapStateToProps = ({ locale }) => ({ locale });

export default connect(mapStateToProps)(DocumentLanguage);
