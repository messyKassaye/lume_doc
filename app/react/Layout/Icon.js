import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { Icon as UIIcon, CountryList } from 'UI';

function getFaSize(size) {
  switch (size) {
    case 'xs':
      return 'sm';
    case 'sm':
      return 'lg';
    case 'md':
      return '2x';
    case 'lg':
      return '2x';
    case 'xlg':
      return '2x';
    default:
      return 'sm';
  }
}

export class Icon extends Component {
  getJSData() {
    const { data } = this.props;
    let _data = data;

    if (data && _data.toJS) {
      _data = _data.toJS();
    }

    return _data;
  }

  getIcon(_data) {
    const { className, size } = this.props;
    let icon;

    if (_data.type === 'Icons') {
      icon = <UIIcon icon={`${_data._id}`} size={`${getFaSize(size)}`} />;
    }

    if (_data.type === 'Flags') {
      const flagClassName = `flag-icon flag-icon-${CountryList.get(_data._id).cca2}`.toLowerCase();
      const componentClassName = `${className} ${flagClassName}`;
      icon = <span className={componentClassName} />;
    }

    return icon;
  }

  render() {
    const { className } = this.props;
    let html = null;
    const _data = this.getJSData();

    if (_data && _data._id) {
      const icon = this.getIcon(_data);
      html = <span className={className}>{icon}</span>;
    }

    return html;
  }
}

Icon.defaultProps = {
  data: undefined,
  className: undefined,
  size: undefined,
};

Icon.propTypes = {
  data: PropTypes.instanceOf(Object),
  className: PropTypes.string,
  size: PropTypes.string,
};

export default connect()(Icon);
