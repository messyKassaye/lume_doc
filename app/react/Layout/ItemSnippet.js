import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React from 'react';

import SafeHTML from 'app/utils/SafeHTML';
import getFieldLabel from 'app/Templates/utils/getFieldLabel';

import t from '../I18N/t';

export const ItemSnippet = ({ snippets, onSnippetClick, template }) => {
  let content;
  let source;
  if (snippets.metadata.length) {
    source = getFieldLabel(snippets.metadata[0].field, template);
    [content] = snippets.metadata[0].texts;
  } else {
    source = t('System', 'Document contents');
    content = snippets.fullText[0].text;
  }
  /* eslint-disable react/no-danger */
  const snippetElement = (
    <>
      <div className="item-snippet-source">{source}</div>
      <div onClick={onSnippetClick} className="item-snippet">
        <SafeHTML>{content}</SafeHTML>
      </div>
    </>
  );

  if (snippets.count === 1) {
    return <div className="item-snippet-wrapper">{snippetElement}</div>;
  }

  return (
    <div className="item-snippet-wrapper">
      {snippetElement}
      <div>
        <a onClick={onSnippetClick}>{t('System', 'Show more')}</a>
      </div>
    </div>
  );
};

ItemSnippet.propTypes = {
  snippets: PropTypes.shape({
    count: PropTypes.number,
    metadata: PropTypes.array,
    fullText: PropTypes.array,
  }).isRequired,
  onSnippetClick: PropTypes.func,
  doc: PropTypes.shape({
    template: PropTypes.string,
  }).isRequired,
  template: PropTypes.shape({
    get: PropTypes.func,
  }).isRequired,
};

ItemSnippet.defaultProps = {
  onSnippetClick: undefined,
};

export const mapStateToProps = (state, ownProps) => ({
  template: state.templates.find(tmpl => tmpl.get('_id') === ownProps.doc.template),
});

export default connect(mapStateToProps)(ItemSnippet);
