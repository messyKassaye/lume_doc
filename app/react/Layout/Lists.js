// TEST!!! Entire file is untested
import PropTypes from 'prop-types';
import React from 'react';
import { Icon } from 'UI';

const List = ({ children }) => <div className="item-group">{children}</div>;

const ItemName = ({ children }) => (
  <div className="item-name">
    <span>{children}</span>
  </div>
);

const ItemFooter = ({ children, onClick }) => (
  <div className="item-actions" onClick={onClick}>
    {children}
  </div>
);

const ProgressBar = ({ progress }) => {
  let message = `${progress} % Completed`;
  let icon = 'upload';
  if (progress === 100) {
    message = 'Processing...';
    icon = 'clock';
  }
  return (
    <div className="label-progress">
      <span className="label label-info">
        <Icon icon={icon} /> <span>{message}</span>
      </span>
      <div className="progress">
        <div
          className="progress-bar progress-bar-striped active"
          style={{ width: `${progress}%` }}
        />
      </div>
    </div>
  );
};

const ItemLabel = ({ children, status }) => {
  let icon = '';
  if (status === 'success') {
    icon = <Icon icon="check" />;
  }
  if (status === 'danger') {
    icon = <Icon icon="times" />;
  }
  if (status === 'warning') {
    icon = <Icon icon="exclamation-triangle" />;
  }
  return (
    <span className={`label label-${status || 'default'}`}>
      {icon} <span>{children}</span>
    </span>
  );
};

ItemFooter.Label = ItemLabel;
ItemFooter.ProgressBar = ProgressBar;

const RowList = ({ children, zoomLevel }) => (
  <div className={`item-group item-group-zoom-${zoomLevel}`}>{children}</div>
);

const RowListItem = ({
  children,
  status,
  onClick,
  onMouseEnter,
  onMouseLeave,
  active,
  className,
}) => {
  let activeClass = '';
  if (active === true) {
    activeClass = 'is-active';
  }
  if (active === false) {
    activeClass = 'is-disabled';
  }

  return (
    <div
      className={`${className} item item-status item-${status || 'default'} ${activeClass}`}
      onClick={onClick}
      onMouseEnter={onMouseEnter}
      onMouseLeave={onMouseLeave}
    >
      {children}
    </div>
  );
};

RowList.Item = RowListItem;

const childrenType = PropTypes.oneOfType([PropTypes.object, PropTypes.array, PropTypes.string]);

List.propTypes = { children: childrenType };

RowList.defaultProps = {
  children: '',
  zoomLevel: 0,
};

RowList.propTypes = {
  children: childrenType,
  zoomLevel: PropTypes.number,
};

RowListItem.propTypes = {
  children: childrenType,
  status: PropTypes.string,
  onClick: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  active: PropTypes.bool,
  className: PropTypes.string,
};
ItemFooter.propTypes = { children: childrenType, onClick: PropTypes.func };
ItemLabel.propTypes = { children: childrenType, status: PropTypes.string };
ItemName.propTypes = { children: childrenType };
ProgressBar.propTypes = { children: childrenType, progress: PropTypes.number };

export { List, RowList, ItemFooter, ItemName };
