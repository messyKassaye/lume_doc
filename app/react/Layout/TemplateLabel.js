import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';
import { createSelector } from 'reselect';
import { COLORS } from 'app/utils/colors';
import t from '../I18N/t';

const getTemplateInfo = createSelector(
  s => s.templates,
  (_s, p) => p.template,
  (templates, currentTemplate) => {
    let styleProps;
    const name = templates.reduce((result, template, index) => {
      if (template.get('_id') === currentTemplate) {
        styleProps = template.get('color')
          ? { className: 'btn-color', style: { backgroundColor: template.get('color') } }
          : { className: `btn-color btn-color-${index % COLORS.length}` };
        return template.get('name');
      }
      return result;
    }, '');

    return { name, styleProps };
  }
);

export class TemplateLabel extends Component {
  render() {
    const { name, template, className, style } = this.props;
    return (
      <span className={className} style={style}>
        {template && t(template, name)}
      </span>
    );
  }
}

TemplateLabel.defaultProps = {
  className: 'btn-color',
  style: undefined,
};

TemplateLabel.propTypes = {
  template: PropTypes.string,
  name: PropTypes.string,
  className: PropTypes.string,
  style: PropTypes.shape({ backgroundColor: PropTypes.string }),
};

const mapStateToProps = (state, props) => {
  const template = getTemplateInfo(state, props);
  return {
    name: template.name,
    template: props.template,
    ...template.styleProps,
  };
};

export default connect(mapStateToProps)(TemplateLabel);
