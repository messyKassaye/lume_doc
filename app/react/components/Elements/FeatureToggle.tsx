import React from 'react';
import { connect } from 'react-redux';
import Immutable from 'immutable';

export type ComponentPropTypes = {
  featureActivated: boolean;
  children: React.ReactNode;
};

export type OwnPropTypes = {
  feature: string;
};

const FeatureToggle: React.FC<ComponentPropTypes> = ({
  featureActivated,
  children,
}: ComponentPropTypes) => (featureActivated ? <>{children}</> : null);

FeatureToggle.defaultProps = {
  featureActivated: false,
};

function mapStateToProps({ settings }: any, ownProps: OwnPropTypes) {
  const features = settings.collection.get('features') || Immutable.fromJS({});
  return {
    featureActivated: features.get(ownProps.feature),
  };
}

const container = connect(mapStateToProps)(FeatureToggle);
export { container as FeatureToggle };
