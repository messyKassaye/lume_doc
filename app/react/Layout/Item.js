import { connect } from 'react-redux';
import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { FormatMetadata } from 'app/Metadata';
import prioritySortingCriteria from 'app/utils/prioritySortingCriteria';

import { FeatureToggle } from 'app/components/Elements/FeatureToggle';
import { FavoriteBanner } from 'app/Favorites';
import helpers from 'app/Documents/helpers';

import { RowList, ItemFooter } from './Lists';
import DocumentLanguage from './DocumentLanguage';
import Icon from './Icon';
import Tip from './Tip';
import ItemSnippet from './ItemSnippet';
import TemplateLabel from './TemplateLabel';

export class Item extends Component {
  getSearchSnipett(doc) {
    if (!doc.snippets || !doc.snippets.count) {
      return false;
    }
    return (
      <ItemSnippet onSnippetClick={this.props.onSnippetClick} snippets={doc.snippets} doc={doc} />
    );
  }

  render() {
    const {
      onClick,
      onMouseEnter,
      onMouseLeave,
      active,
      additionalIcon,
      additionalText,
      buttons,
    } = this.props;

    const doc = helpers.performantDocToJSWithoutRelations(this.props.doc);
    const Snippet = additionalText ? (
      <div className="item-snippet-wrapper">
        <div className="item-snippet">{additionalText}</div>
      </div>
    ) : null;
    const baseClasName = `item-document template-${doc.template}`;
    const itemClassName = `${baseClasName} ${this.props.className || ''}`;
    const itemProps = {
      className: itemClassName,
      onClick,
      onMouseEnter,
      onMouseLeave,
      active,
      tabIndex: '1',
    };

    return (
      <RowList.Item {...itemProps}>
        {this.props.itemHeader}
        <div className="item-info">
          <div className="item-name">
            {additionalIcon || ''}
            <Icon className="item-icon item-icon-center" data={doc.icon} />
            <span>{doc[this.props.titleProperty]}</span>
            <DocumentLanguage doc={this.props.doc} />
          </div>
          {Snippet}
          {this.getSearchSnipett(doc)}
        </div>
        <div className="item-metadata">
          <FormatMetadata
            entity={this.props.noMetadata ? {} : doc}
            sortedProperty={this.props.search.sort}
            additionalMetadata={this.props.additionalMetadata}
            renderLabel={(prop, label) => !prop.noLabel && label}
          />
        </div>
        <ItemFooter>
          {doc.template ? <TemplateLabel template={doc.template} /> : false}
          {doc.published ? '' : <Tip icon="eye-slash">This entity is not public.</Tip>}
          {this.props.labels}
          {buttons}
        </ItemFooter>
        <FeatureToggle feature="favorites">
          <div className="item-favorite">
            <FavoriteBanner sharedId={doc.sharedId} />
          </div>
        </FeatureToggle>
      </RowList.Item>
    );
  }
}

Item.defaultProps = {
  onClick: () => {},
  onMouseEnter: () => {},
  onMouseLeave: () => {},
  noMetadata: false,
};

Item.propTypes = {
  templates: PropTypes.object,
  thesauris: PropTypes.object,
  search: PropTypes.object,
  onClick: PropTypes.func,
  onSnippetClick: PropTypes.func,
  onMouseEnter: PropTypes.func,
  onMouseLeave: PropTypes.func,
  active: PropTypes.bool,
  noMetadata: PropTypes.bool,
  additionalIcon: PropTypes.object,
  additionalText: PropTypes.oneOfType([PropTypes.string, PropTypes.object]),
  additionalMetadata: PropTypes.array,
  doc: PropTypes.object,
  itemHeader: PropTypes.object,
  buttons: PropTypes.object,
  labels: PropTypes.object,
  className: PropTypes.string,
  titleProperty: PropTypes.string,
  evalPublished: PropTypes.bool,
};

Item.defaultProps = {
  search: prioritySortingCriteria.get(),
  titleProperty: 'title',
};

export const mapStateToProps = ({ templates, thesauris }, ownProps) => {
  const search = ownProps.searchParams;
  const _templates = ownProps.templates || templates;
  return { templates: _templates, thesauris, search };
};

export default connect(mapStateToProps)(Item);
