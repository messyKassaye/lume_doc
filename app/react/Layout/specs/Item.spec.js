import { fromJS as Immutable } from 'immutable';
import React from 'react';
import { shallow } from 'enzyme';

import prioritySortingCriteria from 'app/utils/prioritySortingCriteria';
import { FeatureToggle } from 'app/components/Elements/FeatureToggle';
import { FavoriteBanner } from 'app/Favorites';

import { FormatMetadata } from '../../Metadata';
import { Item, mapStateToProps } from '../Item';
import { RowList, ItemFooter } from '../Lists';
import DocumentLanguage from '../DocumentLanguage';
import * as Icon from '../Icon';
import TemplateLabel from '../TemplateLabel';
import ItemSnippet from '../ItemSnippet';

describe('Item', () => {
  let component;
  let props;

  beforeEach(() => {
    Icon.default = Icon.Icon;
    props = {
      doc: Immutable({
        type: 'entity',
        icon: { _id: 'icon', type: 'Icons' },
        title: 'doc title',
        template: 'templateId',
        sharedId: 'sharedId',
        creationDate: 123,
        snippets: [],
      }),
      active: true,
      className: 'custom-class',
      onClick: jasmine.createSpy('onClick'),
      onSnippetClick: jasmine.createSpy('onSnippetClick'),
      onMouseEnter: jasmine.createSpy('onMouseEnter'),
      onMouseLeave: jasmine.createSpy('onMouseLeave'),
      additionalIcon: <div>additionalIcon</div>,
      buttons: <div>Buttons</div>,
      templates: Immutable([]),
      thesauris: Immutable([]),
    };
  });

  const render = () => {
    component = shallow(<Item {...props} />);
  };

  it('should have default props values assigned', () => {
    render();
    expect(component.instance().props.search).toEqual(prioritySortingCriteria.get());
  });

  it('should extend RowList.Item and append active, type and classNames correctly', () => {
    render();
    expect(component.find(RowList.Item).props().className).toContain('item-document');
    expect(component.find(RowList.Item).props().className).toContain('template-templateId');
    expect(component.find(RowList.Item).props().className).toContain('custom-class');
    expect(component.find(RowList.Item).props().active).toBe(true);
  });

  it('should not fail on legacy entities without template', () => {
    props.doc = props.doc.set('template', undefined);
    render();
    expect(component.find(RowList.Item).props().className).toContain('template-undefined');
  });

  it('should replicate onClick, onMouseEnter and onMouseLeave behaviours of parent', () => {
    render();
    component.find(RowList.Item).simulate('click');
    expect(props.onClick).toHaveBeenCalled();

    component.find(RowList.Item).simulate('mouseEnter');
    expect(props.onMouseEnter).toHaveBeenCalled();

    component.find(RowList.Item).simulate('mouseLeave');
    expect(props.onMouseLeave).toHaveBeenCalled();
  });

  it('should include a header if present', () => {
    props.itemHeader = <div className="item-header">Item Header</div>;
    render();

    expect(component.find('.item-header').text()).toBe('Item Header');
  });

  it('should include additionalIcon, icon and title in the components name', () => {
    render();
    expect(component.find('.item-name').text()).toContain('additionalIcon');
    expect(component.find('.item-name').text()).toContain('doc title');
    expect(
      component
        .find('.item-name')
        .find(Icon.default)
        .props().data
    ).toEqual({ _id: 'icon', type: 'Icons' });
    expect(
      component
        .find('.item-name')
        .find(DocumentLanguage)
        .props().doc
    ).toBe(props.doc);
  });

  it('should include feature-toggled FavoriteBanner passed the sharedId', () => {
    render();
    const FeatureToggleSection = component.find({ feature: 'favorites' });
    const FavoriteBannerSection = FeatureToggleSection.childAt(0).childAt(0);

    expect(FeatureToggleSection.type()).toBe(FeatureToggle);
    expect(FeatureToggleSection.childAt(0).props().className).toContain('item-favorite');
    expect(FavoriteBannerSection.type()).toBe(FavoriteBanner);
    expect(FavoriteBannerSection.props().sharedId).toBe('sharedId');
  });

  it('should accept a different property name for the title', () => {
    props.doc = props.doc.set('label', 'label as title');
    props.titleProperty = 'label';
    render();
    expect(component.find('.item-name').text()).toContain('label as title');
  });

  it('should include a template label and custom buttons inside the footer', () => {
    render();
    expect(
      component
        .find(ItemFooter)
        .find(TemplateLabel)
        .props().template
    ).toBe('templateId');
    expect(
      component
        .find(ItemFooter)
        .find('div')
        .at(0)
        .text()
    ).toContain('Buttons');
  });

  describe('Metadata', () => {
    it('should render FormatMetadata passing entity sort property and additionalMetadata', () => {
      props.search = { sort: 'sortedProperty' };
      props.additionalMetadata = ['additioal', 'metadata'];
      render();
      expect(component.find(FormatMetadata).props().entity).toEqual(props.doc.toJS());
      expect(component.find(FormatMetadata).props().sortedProperty).toBe(props.search.sort);
      expect(component.find(FormatMetadata).props().additionalMetadata).toBe(
        props.additionalMetadata
      );
    });
  });

  describe('when doc have no snippets', () => {
    it('should not render snippet secction when undefined', () => {
      props.doc = Immutable({
        type: 'entity',
        icon: { _id: 'icon', type: 'Icons' },
        title: 'doc title',
        template: 'templateId',
        creationDate: 123,
      });

      render();
      expect(component.find('.item-snippet').length).toBe(0);
    });
    it('should not render snippet secction when empty', () => {
      props.doc = Immutable({
        type: 'entity',
        icon: { _id: 'icon', type: 'Icons' },
        title: 'doc title',
        template: 'templateId',
        creationDate: 123,
        snippets: [],
      });

      render();
      expect(component.find('.item-snippet').length).toBe(0);
    });
  });

  describe('when doc have snippets', () => {
    it('should render ItemSnippet including doc, snippets and onSnippetClick props', () => {
      props.doc = Immutable({
        type: 'entity',
        icon: { _id: 'icon', type: 'Icons' },
        title: 'doc title',
        template: 'templateId',
        creationDate: 123,
        snippets: {
          count: 1,
          metadata: [],
          fullText: [{ text: '<span>snippet!</span>', page: 1 }],
        },
      });

      render();
      expect(component.find(ItemSnippet)).toMatchSnapshot();
    });
  });

  describe('mapStateToProps', () => {
    let templates;
    let thesauris;
    let search;

    beforeEach(() => {
      templates = 'templates';
      thesauris = 'thesauris';
    });

    it('should include templates, thesauris and default sort', () => {
      expect(mapStateToProps({ templates, thesauris }, {})).toEqual({
        templates,
        thesauris,
        search,
      });
    });

    it('should allow overriding the default sort', () => {
      const ownProps = { searchParams: { sort: 'newSort' } };
      expect(mapStateToProps({ templates, thesauris }, ownProps)).toEqual({
        templates,
        thesauris,
        search: { sort: 'newSort' },
      });
    });
  });
});
