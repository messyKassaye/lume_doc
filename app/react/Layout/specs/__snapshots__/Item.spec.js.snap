// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Item when doc have snippets should render ItemSnippet including doc, snippets and onSnippetClick props 1`] = `
<Connect(ItemSnippet)
  doc={
    Object {
      "creationDate": 123,
      "icon": Object {
        "_id": "icon",
        "type": "Icons",
      },
      "snippets": Object {
        "count": 1,
        "fullText": Array [
          Object {
            "page": 1,
            "text": "<span>snippet!</span>",
          },
        ],
        "metadata": Array [],
      },
      "template": "templateId",
      "title": "doc title",
      "type": "entity",
    }
  }
  onSnippetClick={[Function]}
  snippets={
    Object {
      "count": 1,
      "fullText": Array [
        Object {
          "page": 1,
          "text": "<span>snippet!</span>",
        },
      ],
      "metadata": Array [],
    }
  }
/>
`;
